File: src/claw_mem/dreaming/pipeline.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Dreaming Engine — Pipeline (v4.12.0)

Orchestrates the complete light→deep→REM→promote pipeline.
Supports dry_run mode (score-only, no persistence).
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .config import DreamingConfig
    from .rem import REMResult
    from .promote import PromotionResult


@dataclass
class DreamingResult:
    """Full result of a dreaming pipeline run.

    Attributes:
        staged: Number of signals staged in light phase.
        scored: Number of candidates scored in deep phase.
        passed: Number of candidates passing the filter.
        promoted: Total promotions (episodic + semantic + procedural).
        skills_stored: Number of skills persisted.
        duration_ms: Pipeline wall-clock duration in milliseconds.
        dry_run: Whether this was a dry run.
        error: Error message if the pipeline failed (None on success).
        promotion_detail: Detailed PromotionResult (None on dry_run).
    """

    staged: int = 0
    scored: int = 0
    passed: int = 0
    promoted: int = 0
    skills_stored: int = 0
    duration_ms: float = 0.0
    dry_run: bool = False
    error: Optional[str] = None
    promotion_detail: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        d = {
            "staged": self.staged,
            "scored": self.scored,
            "passed": self.passed,
            "promoted": self.promoted,
            "skills_stored": self.skills_stored,
            "duration_ms": round(self.duration_ms, 2),
            "dry_run": self.dry_run,
        }
        if self.error:
            d["error"] = self.error
        if self.promotion_detail:
            d["promotion_detail"] = self.promotion_detail
        return d


class DreamingPipeline:
    """Full dreaming pipeline: light → deep → REM → promote.

    Usage::

        from claw_mem import MemoryManager
        from claw_mem.dreaming import DreamingPipeline, DreamingConfig

        mm = MemoryManager()
        config = DreamingConfig(dry_run=True)
        pipeline = DreamingPipeline(mm, config=config)
        result = pipeline.run()
    """

    def __init__(
        self,
        memory_manager: Any,
        config: Optional[DreamingConfig] = None,
        llm_provider: Any = None,
    ):
        from .config import DreamingConfig

        self._mm = memory_manager
        self._config = config or DreamingConfig()
        self._llm = llm_provider

        # Last run stats
        self._last_result: Optional[DreamingResult] = None

    # ── public API ─────────────────────────────────────────────────

    def run(self) -> DreamingResult:
        """Execute the full dreaming pipeline.

        Returns:
            DreamingResult with stage counts and timing.
        """
        t0 = time.time()
        result = DreamingResult(dry_run=self._config.dry_run)

        try:
            # ── Phase 1: Light (ingest signals) ────
            from .light import SignalIngestor

            ingestor = SignalIngestor(self._mm, config=self._config)
            staged_count = ingestor.ingest()
            result.staged = staged_count

            if staged_count == 0:
                result.duration_ms = (time.time() - t0) * 1000
                self._last_result = result
                return result

            signals = ingestor._staged

            # ── Phase 2: Deep (score candidates) ────
            from .deep import CandidateScorer

            scorer = CandidateScorer(config=self._config)
            all_candidates = scorer.score_all(signals)
            result.scored = len(all_candidates)

            passed = scorer.filter(all_candidates)
            result.passed = len(passed)

            if len(passed) == 0:
                result.duration_ms = (time.time() - t0) * 1000
                self._last_result = result
                return result

            # ── Phase 3: REM (extract patterns) ────
            from .rem import PatternExtractor

            extractor = PatternExtractor(llm_provider=self._llm)
            rem_result = extractor.extract(passed)

            # ── Phase 4: Promote (persist) ────
            from .promote import Promoter

            promoter = Promoter(self._mm, dry_run=self._config.dry_run)
            promotion = promoter.promote(passed, rem_result)
            result.promoted = promotion.total
            result.skills_stored = promotion.skill_stored
            result.promotion_detail = promotion.to_dict()

            # ── Store skills from REM ────
            if rem_result.skills and not self._config.dry_run:
                result.skills_stored = promotion.skill_stored

        except Exception as e:
            result.error = str(e)

        result.duration_ms = (time.time() - t0) * 1000
        self._last_result = result
        return result

    def last_result(self) -> Optional[DreamingResult]:
        """Get the result of the last pipeline run."""
        return self._last_result

File: src/dreaming/pipeline.ts
// Copyright 2026 Peter Cheng
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Dreaming Engine -- Pipeline (v4.12.0)
 *
 * Orchestrates the complete light->deep->REM->promote pipeline.
 * Supports dry_run mode (score-only, no persistence).
 */

import { DEFAULT_DREAMING_CONFIG, type DreamingConfig } from "./config";
import { SignalIngestor, type MemoryManagerLike, type Signal } from "./light";
import { CandidateScorer, type ScoredCandidate } from "./deep";
import { PatternExtractor, type REMResult } from "./rem";
import { Promoter, type MemoryManagerWithStorage, type PromotionResult } from "./promote";

// ── DreamingResult ───────────────────────────────────────────────────────

export interface DreamingResult {
  /** Number of signals staged in light phase. */
  staged: number;
  /** Number of candidates scored in deep phase. */
  scored: number;
  /** Number of candidates passing the filter. */
  passed: number;
  /** Total promotions (episodic + semantic + procedural). */
  promoted: number;
  /** Number of skills persisted. */
  skillsStored: number;
  /** Pipeline wall-clock duration in milliseconds. */
  durationMs: number;
  /** Whether this was a dry run. */
  dryRun: boolean;
  /** Error message if the pipeline failed (undefined on success). */
  error?: string;
  /** Detailed PromotionResult (undefined on dry_run). */
  promotionDetail?: Record<string, unknown>;
}

export function createDreamingResult(overrides?: Partial<DreamingResult>): DreamingResult {
  return {
    staged: 0,
    scored: 0,
    passed: 0,
    promoted: 0,
    skillsStored: 0,
    durationMs: 0,
    dryRun: false,
    ...overrides,
  };
}

export function dreamingResultToDict(r: DreamingResult): Record<string, unknown> {
  const d: Record<string, unknown> = {
    staged: r.staged,
    scored: r.scored,
    passed: r.passed,
    promoted: r.promoted,
    skills_stored: r.skillsStored,
    duration_ms: Math.round(r.durationMs * 100) / 100,
    dry_run: r.dryRun,
  };
  if (r.error) d.error = r.error;
  if (r.promotionDetail) d.promotion_detail = r.promotionDetail;
  return d;
}

// Combine both memory manager interfaces for pipeline usage
type FullMemoryManager = MemoryManagerLike & MemoryManagerWithStorage;

// ── DreamingPipeline ─────────────────────────────────────────────────────

export class DreamingPipeline {
  private _mm: FullMemoryManager;
  private _config: DreamingConfig;
  private _llm: unknown;
  private _lastResult: DreamingResult | undefined;

  constructor(
    memoryManager: FullMemoryManager,
    config?: DreamingConfig,
    llmProvider?: unknown,
  ) {
    this._mm = memoryManager;
    this._config = config ?? DEFAULT_DREAMING_CONFIG;
    this._llm = llmProvider;
  }

  /**
   * Execute the full dreaming pipeline.
   *
   * @returns DreamingResult with stage counts and timing.
   */
  run(): DreamingResult {
    const t0 = Date.now();
    const result = createDreamingResult({ dryRun: this._config.dryRun });

    try {
      // ── Phase 1: Light (ingest signals) ────
      const ingestor = new SignalIngestor(this._mm, this._config);
      const stagedCount = ingestor.ingest();
      result.staged = stagedCount;

      if (stagedCount === 0) {
        result.durationMs = Date.now() - t0;
        this._lastResult = result;
        return result;
      }

      const signals: Signal[] = ingestor["_internalStaged"];

      // ── Phase 2: Deep (score candidates) ────
      const scorer = new CandidateScorer(this._config);
      const allCandidates = scorer.scoreAll(signals);
      result.scored = allCandidates.length;

      const passed = scorer.filter(allCandidates);
      result.passed = passed.length;

      if (passed.length === 0) {
        result.durationMs = Date.now() - t0;
        this._lastResult = result;
        return result;
      }

      // ── Phase 3: REM (extract patterns) ────
      const extractor = new PatternExtractor(this._llm);
      const remResult: REMResult = extractor.extract(passed);

      // ── Phase 4: Promote (persist) ────
      const promoter = new Promoter(this._mm, this._config.dryRun);
      const promotion: PromotionResult = promoter.promote(passed, remResult);
      result.promoted =
        promotion.episodicPromoted +
        promotion.semanticReinforced +
        promotion.proceduralPromoted;
      result.skillsStored = promotion.skillStored;
      result.promotionDetail = {
        episodic_promoted: promotion.episodicPromoted,
        semantic_reinforced: promotion.semanticReinforced,
        procedural_promoted: promotion.proceduralPromoted,
        skill_stored: promotion.skillStored,
        total: result.promoted,
        dry_run: promotion.dryRun,
      };

      // Store skills from REM
      if (remResult.skills.length > 0 && !this._config.dryRun) {
        result.skillsStored = promotion.skillStored;
      }
    } catch (e) {
      result.error = e instanceof Error ? e.message : String(e);
    }

    result.durationMs = Date.now() - t0;
    this._lastResult = result;
    return result;
  }

  /** Get the result of the last pipeline run. */
  lastResult(): DreamingResult | undefined {
    return this._lastResult;
  }
}
